File: src/main/java/com/dansoftware/libraryapp/util/OsInfo.java
package com.dansoftware.libraryapp.util;

import org.jetbrains.annotations.NotNull;
import oshi.PlatformEnum;
import oshi.SystemInfo;
import oshi.software.os.OperatingSystem;

/**
 * Provides utility methods for detecting the type of the particular operating system.
 *
 * @author Daniel Gyorffy
 */
public final class OsInfo {
    private OsInfo() {
    }

    public static boolean is(@NotNull PlatformEnum platformType,
                             @NotNull String versionStartsWith) {
        SystemInfo systemInfo = new SystemInfo();
        OperatingSystem operatingSystem = systemInfo.getOperatingSystem();
        OperatingSystem.OSVersionInfo versionInfo = operatingSystem.getVersionInfo();
        return OsInfo.is(platformType) && versionInfo.getVersion().startsWith(versionStartsWith);
    }

    public static boolean is(@NotNull PlatformEnum platformType) {
        return SystemInfo.getCurrentPlatformEnum().equals(platformType);
    }

    public static boolean isWindows() {
        return is(PlatformEnum.WINDOWS);
    }

    public static boolean isLinux() {
        return is(PlatformEnum.LINUX);
    }

    public static boolean isMac() {
        return is(PlatformEnum.MACOSX);
    }

    public static boolean isWindows10() {
        return is(PlatformEnum.WINDOWS, "10");
    }
}

File: src/main/java/com/dansoftware/libraryapp/util/OsInfo.kt
package com.dansoftware.libraryapp.util

import oshi.PlatformEnum
import oshi.SystemInfo

/**
 * Provides utility methods for detecting the type of the particular operating system.
 *
 * @author Daniel Gyorffy
 */
@Suppress("unused")
object OsInfo {
    private val platformType: PlatformEnum
    private val version: String
    private val buildNumber: String
    private val name: String

    init {
        val operatingSystem = SystemInfo().operatingSystem
        val osVersionInfo = operatingSystem.versionInfo

        this.platformType = SystemInfo.getCurrentPlatformEnum()
        this.version = osVersionInfo.version
        this.buildNumber = osVersionInfo.buildNumber
        this.name = operatingSystem.family
    }

    @JvmStatic
    fun isWindows(): Boolean = hasType(PlatformEnum.WINDOWS)

    @JvmStatic
    fun isLinux(): Boolean = hasType(PlatformEnum.LINUX)

    @JvmStatic
    fun isMac(): Boolean = hasType(PlatformEnum.MACOSX)

    @JvmStatic
    fun isWindows10(): Boolean = hasTypeAndVersion(PlatformEnum.WINDOWS, "10")

    @JvmStatic
    fun isWindows10OrLater(): Boolean = hasTypeAndVersionOrHigher(PlatformEnum.WINDOWS, "10")

    @JvmStatic
    fun isMacOsMojaveOrLater(): Boolean = hasTypeAndVersionOrHigher(PlatformEnum.MACOSX, "10.14")

    @JvmStatic
    fun getName(): String = name

    @JvmStatic
    fun getVersion(): String = version

    @JvmStatic
    fun getBuildNumber(): String = buildNumber

    @JvmStatic
    fun getPlatformVersion(): String = version

    @JvmStatic
    fun getPlatformType(): PlatformEnum = platformType

    @JvmStatic
    fun hasTypeAndVersion(platformType: PlatformEnum, versionStarts: String): Boolean =
            hasType(platformType) && this.version.startsWith(versionStarts)

    @JvmStatic
    fun hasTypeAndVersionOrHigher(platformType: PlatformEnum, version: String): Boolean =
        hasType(platformType) && hasVersionOrHigher(version)

    @JvmStatic
    fun hasType(platformType: PlatformEnum): Boolean = this.platformType == platformType

    @JvmStatic
    fun hasVersionOrHigher(version: String) =
        this.version.replace(".", "").toInt() >= version.replace(".", "").toInt()
}